---
title: "Maize RNA-seq report"
author: "Qizo Zhenghao"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
    toc: yes
    toc_float:
      collapsed: no
    keep_md: false
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
  file_name: ""
  species_names: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(knitr))
knitr::opts_chunk$set(fig.width = 10,
                      fig.height = 6,
                      fig.path = "Figure",
                      echo = FALSE,
                      cache = FALSE,
                      prompt = FALSE,
                      tidy = FALSE,
                      comment = NA,
                      message = FALSE,
                      warning = FALSE,
                      eval = T,
                      rownames.print = FALSE)
```

```{r}
options(stringsAsFactors = F)
library(patchwork)
library(gmodels)
library(dplyr)
library(tidyverse)
library(reshape2)
library(ggrepel)
library(knitr)
library(DESeq2)
library(corrplot)
library(tidyr)
library(glue)
library(ggpubr)
library(edgeR)
library(kableExtra)
library(ggplot2)
library(clusterProfiler)
library(AnnotationHub)
library(ggthemes)
library(biomaRt)
library(DT)
```

```{r loading data}
species_names <- params$species_names
if(species_names=="mouse") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/mm10/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "")
  } else if(species_names=="pig") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/ss11/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "")
  } else if(species_names=="hamster") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/MesAur1.0/genome_annotation_data/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "")
  } else if(species_names=="cow") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/ARS-UCD1.2/genome_annotation_data/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "")
  } else if(species_names=="dm6") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/Drosophila_Ref/Flybase_id_trans.txt",header = T,sep = "\\",quote = "")
  } else if(species_names=="human") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/hg38/genome_annotation_data/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "",fill = TRUE)
  } else if(species_names=="Zebrafish") {
    ID_trans <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/GRCz11/genome_annotation_data/Ensemble_id_trans.txt",header = T,sep = "\\",quote = "",fill = TRUE)
  } else {
    print("No truth found")
}

input_dir <- params$input_dir
input_dir <- normalizePath(input_dir)
file_name <- params$file_name
reads_table = read.table(paste0(input_dir,"/","01_merge_reads.txt"),header = T,sep = "\t")
counts_table = read.table(paste0(input_dir,"/","02-megre_counts.txt"),header = T,sep = "\t")
gene_exon <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/Maize/B73_gene_exon_length.txt",header = T,sep = "\t")
counts_table <- counts_table[,c(1:ncol(counts_table)-1)]

# read counts 
# reads_table = read.table("/data/yulab/qzh/Analysis/qzh/tmp/03-R4//Summary/01_merge_reads.txt",header = T,sep = "\t")
# counts_table = read.table("/data/yulab/qzh/Analysis/qzh/tmp/03-R4//Summary/02-megre_counts.txt",header = T,sep = "\t")
# gene_exon <- read.table("/data/yulab/qzh/Analysis/qzh/data_Ref/Maize/B73_gene_exon_length.txt",header = T,sep = "\t")
# counts_table <- counts_table[,c(1:ncol(counts_table)-1)]

```

```{r proces data}
clean_table_1 <- reads_table[,c(1:4)]
clean_table_1$short_reads <- clean_table_1$raw_reads-clean_table_1$clean_reads
clean_table_1$short_pct <- round(clean_table_1$short*100/clean_table_1$raw_reads,2)
clean_table_1$clean_pct <- round(clean_table_1$clean_reads*100/clean_table_1$raw_reads,2)

clean_table_2<-tidyr::gather(
  data=clean_table_1,
  key="group",
  value="reads",
  clean_reads:short_reads
)

map_table_1 <- reads_table[,c(1,5:7)]
names(map_table_1) <- c("sample","rRNA","input","map")
map_table_1$unmap <- map_table_1$input-map_table_1$map
map_table_1$map_pct <- round(map_table_1$map*100/map_table_1$input,2)
map_table_2<-tidyr::gather(
  data=map_table_1[,c(1,2,4,5)],
  key="group",
  value="reads",
  rRNA:unmap
)

if (nrow(reads_table) <=2 ) {
  # edgeR
  condition <- factor(reads_table[,2],levels = c("WT","Mut"))
  gene_table <- DGEList(counts = counts_table[,c(2:3)],genes = counts_table[,1],group = condition)
  gene_table_high <- rowSums(cpm(gene_table)>1) >= 1
  gene_table <- gene_table[gene_table_high, , keep.lib.sizes=FALSE]
  gene_table <- calcNormFactors(gene_table)
  # RPKM
  gene_counts <- cbind(gene_table$genes,gene_table$counts)
  overlap_table <- inner_join(gene_counts, gene_exon, by="genes")
  rownames(gene_counts) <- gene_counts[,1]
  gene_counts<-gene_counts[,-1]
  kb <- overlap_table$length / 1000
  countdata <- gene_counts[,1:ncol(gene_counts)]
  rpk <- countdata / kb
  fpkm <- data.frame(t(t(rpk)/colSums(countdata) * 10^6))
  names(fpkm) <- reads_table$group
  fpkm$baseMean = apply(fpkm,1,mean) 
  fpkm$gene <- row.names(fpkm)
  
  # FDR and pvalue
  y_bcv <- gene_table
  bcv <- 0.3
  et <- exactTest(y_bcv, dispersion = bcv ^ 2)
  DEG <- cbind(gene_table$genes,fpkm$baseMean,et$table)
  DEG <- DEG[,-4]
  names(DEG) <- c("gene","baseMean","log2FoldChange","pvalue")
  DEG$change = as.factor(ifelse(abs(DEG$log2FoldChange) >=1 & DEG$pvalue <= 0.05,
                                ifelse(DEG$log2FoldChange >= 1,'UP','DOWN'),'NOT'))
  DEG_gene_table <- cbind(DEG,fpkm[,c(1:2)])
} else {
  # DEseq2
  exprSet <- counts_table
  rownames(exprSet) <- exprSet[,1]
  exprSet <- exprSet[,-1]
  condition <- factor(reads_table[,2],levels = c("WT","Mut"))
  colData <- data.frame(sampleName = colnames(exprSet),
                        fileName = colnames(exprSet),
                        condition = condition)

  dds <- DESeqDataSetFromMatrix(exprSet,colData,design= ~condition)
  dds <- DESeq(dds)
  # resultsNames(dds)
  resLFC <- lfcShrink(dds, coef="condition_Mut_vs_WT", type="apeglm")
  DEG <- as.data.frame(resLFC)
  DEG$gene <- row.names(DEG)
  DEG <- na.omit(DEG)
  DEG$change = as.factor(ifelse(abs(DEG$log2FoldChange) >=1 & DEG$pvalue <= 0.05,
                              ifelse(DEG$log2FoldChange >= 1,'UP','DOWN'),'NOT'))

  normalized_counts <- data.frame(counts(dds, normalized=TRUE))
  normalized_counts$gene <- rownames(normalized_counts)
  group_WT <- reads_table[which(reads_table$group=="WT"),1]
  group_Mut <- reads_table[which(reads_table$group=="Mut"),1]
  normalized_counts$WT <- apply(normalized_counts[,which(colnames(normalized_counts) %in% group_WT)],1,mean)
  normalized_counts$Mut <- apply(normalized_counts[,which(colnames(normalized_counts) %in% group_Mut)],1,mean)
  normalized_counts_mean <- normalized_counts[,c("gene","WT","Mut")]
  DEG_gene_table <- merge(DEG,normalized_counts_mean,by.x="gene",by.y="gene")
  DEG_gene_table <- DEG_gene_table[,-c(4,6)]
}

# ensembl = useMart(biomart = "plants_mart",host = "http://plants.ensembl.org")
# dataset <- listDatasets(mart = ensembl)

ensembl = useMart(biomart = "plants_mart",host = "http://plants.ensembl.org",dataset="zmays_eg_gene")
attributes <- listAttributes(ensembl)

ensemble_id <- getBM(attributes =c("ensembl_gene_id","entrezgene_id"),filters = "ensembl_gene_id",values = DEG_gene_table$gene,mart = ensembl)
names(ensemble_id) <- c("gene","ENTREZID")
####### step 3 down database 
hub <- AnnotationHub()
maize <- hub[['AH111691']]
# columns(maize)
gene_name <- bitr(ensemble_id$ENTREZID,fromType = 'ENTREZID',
           toType = c('ENTREZID','SYMBOL','GENENAME'),
           OrgDb=maize,drop = F)
gene_name_entre <- merge(ensemble_id,gene_name,by.x="ENTREZID",by.y="ENTREZID")
DEG_table <- merge(DEG_gene_table,gene_name_entre,by.x="gene",by.y="gene")
write_csv(DEG_table,file = paste0(input_dir,"/","03-Normalzie_gene_counts.csv"))
```

```{r config}
# sample information
n_samples <- nrow(reads_table)
total <- sum(reads_table$raw_reads)/1000000
map <- sum(reads_table$map_reads)/1000000
map_pct <- round(map*100/total,3)
```
## 1 Summary

```{r summary, results="asis"}
sum_text <- glue::glue(
  "A total of {n_samples} samples; {total}M sequencing depth; ",
  "with {map}M ({map_pct}%) reads mapped to reference genome."
)
print(sum_text)
```

## 2. Results

### 2.1 Table 1. Trimming reads
```{r}
DT::datatable(
  clean_table_1,
  extensions   = 'Buttons',
  options      = list(
    pageLength = 10,
    columnDefs = list(list(className = 'dt-center',targets = 0:5)),
    scrollX    = TRUE,
    dom        = 'Bfrtip',
    buttons    =
      list('copy', #'print',
        list(extend  = 'collection',
        buttons = c('excel', 'csv'),
        text    = 'Download')))
)
```


```{r fig.width=6, fig.height=3, fig.align = 'center'}
trim_plot <- ggplot(clean_table_2,aes(x=reads, y=sample,fill=group)) +
  geom_bar(position = "fill",stat="identity") +
  scale_x_continuous(expand = c(0,0)) +
  guides(fill=guide_legend(reverse=TRUE)) +
  labs(title="Trim_reads",x="",y="species") +
  theme_classic() +
  scale_fill_manual(values=c(short_reads = "#791214", clean_reads = "#4ab2b8")) + 
  theme(plot.title = element_text(size=12,hjust=0.5, face = "bold"),
        axis.text.y = element_text(size = 8,vjust = 0.5, hjust = 1,color = "grey45"),
        axis.text.x = element_text(size = 8,vjust = 1,color = "grey45"))
print(trim_plot)
```

### 2.2 Table 2. Mapping reads
```{r}
DT::datatable(
  map_table_1,
  extensions   = 'Buttons',
  options      = list(
    pageLength = 10,
    columnDefs = list(list(className = 'dt-center',targets = 0:5)),
    scrollX    = TRUE,
    dom        = 'Bfrtip',
    buttons    =
      list('copy', #'print',
        list(extend  = 'collection',
        buttons = c('excel', 'csv'),
        text    = 'Download')))
)
```


```{r fig.width=6, fig.height=3, fig.align = 'center'}
map_plot <- ggplot(map_table_2,aes(x=reads, y=sample,fill=group)) +
  geom_bar(position = "fill",stat="identity") +
  scale_x_continuous(expand = c(0,0)) +
  guides(fill=guide_legend(reverse=TRUE)) +
  labs(title="Trim_reads",x="",y="species") +
  theme_classic() +
  scale_fill_manual(values=c(rRNA = "#f6871f", map = "#4ab2b8",unmap="#791214")) + 
  theme(plot.title = element_text(size=12,hjust=0.5, face = "bold"),
        axis.text.y = element_text(size = 8,vjust = 0.5, hjust = 1,color = "grey45"),
        axis.text.x = element_text(size = 8,vjust = 1,color = "grey45"))
print(map_plot)
```


### 2.3 RNAseq library type

Check the RNAseq library type, check if the reads from forward strand or reverse strand of mRNA;

Using `featureCounts` to count reads on mRNAs.

### 2.4 Correlation between replicates

#### 1 Correlation - spearman
```{r fig.width=6, fig.height=6, fig.align = 'center'}
#################
if (nrow(reads_table) <=2 ) {
  normalized_counts <- fpkm[,c(1:2)]
} else {
  normalized_counts <- counts(dds, normalized=TRUE)
}

corr <- cor(normalized_counts, method = 'spearman')
corrplot(corr, type = 'upper', tl.col = 'black', order = 'hclust', tl.srt = 45, addCoef.col = 'white')	
########## calculate PCA###########
```

#### 2 Correlation - PCA
```{r, fig.width=6, fig.height=5, fig.align = 'center'}
pca.info <- fast.prcomp(normalized_counts)
pca.data <- data.frame(sample = rownames(pca.info$rotation),Type = reads_table[,2],pca.info$rotation)
PCA_plot = ggscatter(pca.data,x = "PC1",y = "PC2",color = "Type") + theme_base() 
PCA_plot = PCA_plot + geom_text_repel(data=pca.data,aes(x = PC1,y = PC2,label = sample),size=6)
PCA_plot
```

#### 3 Correlation - person
```{r fig.width=7, fig.height=5, fig.align = 'center'}
dataPearson <- round(cor(normalized_counts,method = c("pearson")) , 2)
data_person <- as.data.frame(dataPearson) %>%
  mutate(x=rownames(dataPearson)) %>%
  melt(id='x')
names(data_person) <- c("x","y","Corr")
#先确定好绘图时x轴、y轴的因子顺序：
list <- rownames(dataPearson)
list <- factor(list,levels = list)

ggplot(data_person,aes(factor(x,levels = list),
                factor(y,levels = list), #定义x，y轴顺序，防止被默认改变
                fill=Corr))+  #根据相关性值填充颜色
  geom_tile(color = "white") +
  geom_text(data=data_person,aes(x = x,y = y,label = Corr),size=6) +
  scale_fill_gradient2(low = "white",high ='#4574b3',
                       limits=c(0,1),breaks=c(0,0.2,0.4,0.6,0.8,1))+
  labs(x=NULL,y=NULL)+
  theme_bw(base_size = 15)
```


## 3 Differentially expressed genes

### 3.1 Info

```{r}
wt_name <- strsplit(reads_table[which(reads_table$group=="WT"),1][1],"_rep")[[1]][1]
mut_name <- strsplit(reads_table[which(reads_table$group=="Mut"),1][1],"_rep")[[1]][1]
msg <- glue::glue(
  "Configuration:",
  " Compare: log2( Mut / WT )",
  "  Genome: B73_v5",
  "wildtype: {wt_name}",
  "  mutant: {mut_name}",
  "quantification: featureCounts (v2.0.0)",
  "DE analysis: replicates：DESeq2 (v1.30.1) single：edgeR",
  "cutoff: padj < 0.05 and |log2fc| >= 1",
  .sep = "\n")
print(msg)
```

### 3.2 Number of differentially expressed genes

Here are the numbers of significant changed genes, criteria: `foldChange >= 2, pvalue < 0.05`

```{r}
diff_group <- c(nrow(filter(DEG_table,DEG_table$change=="UP")),
  nrow(filter(DEG_table,DEG_table$change=="DOWN")),
  nrow(filter(DEG_table,DEG_table$change=="NOT")))

diff_gene_table <- data.frame(diff = c("UP","DOWN","NOT"),
                          diff_gene_numbers = diff_group,
                          shrink = rep("apeglm",3)
                          ,row.names = c(1,2,3))
DT::datatable(
  diff_gene_table,
  extensions   = 'Buttons',
  options      = list(
    pageLength = 10,
    columnDefs = list(list(className = 'dt-center',targets = 0:3)),
    scrollX    = TRUE,
    dom        = 'Bfrtip',
    buttons    =
      list('copy', #'print',
        list(extend  = 'collection',
        buttons = c('excel', 'csv'),
        text    = 'Download')))
)
```

### 3.3 Up-regulated genes

```{r}
## shrink
DEG_table_round <- DEG_table
DEG_table_round[,c(2:4)] <- round(DEG_table[,c(2:4)],3)
DEG_table_round[,c(6,7)] <- round(DEG_table[,c(6,7)],3)

UP_gene_table <- DEG_table_round %>%
  dplyr::select(!c("baseMean")) %>%
  dplyr::filter(change=="UP")

DT::datatable(
  UP_gene_table,
  extensions   = 'Buttons',
  options      = list(
    pageLength = 10,
    columnDefs = list(list(className = 'dt-center',targets = 0:3)),
    scrollX    = TRUE,
    dom        = 'Bfrtip',
    buttons    =
      list('copy', #'print',
        list(extend  = 'collection',
        buttons = c('excel', 'csv'),
        text    = 'Download')))
)
```


### 3.4 Down-regulated genes

```{r}
## shrink
UP_gene_table <- DEG_table_round %>%
  dplyr::select(!c("baseMean")) %>%
  dplyr::filter(change=="DOWN")

DT::datatable(
  UP_gene_table,
  extensions   = 'Buttons',
  options      = list(
    pageLength = 10,
    columnDefs = list(list(className = 'dt-center',targets = 0:3)),
    scrollX    = TRUE,
    dom        = 'Bfrtip',
    buttons    =
      list('copy', #'print',
        list(extend  = 'collection',
        buttons = c('excel', 'csv'),
        text    = 'Download')))
)
```

## 4 DESeq quality control

### 4.4 MA plot
```{r, fig.width=6, fig.height=6, fig.align = 'center'}
this_tile <- paste0('Cutoff for logFC is ',round(1,3),
                    '\nThe number of up gene is ',nrow(DEG[DEG$change == 'UP',]),
                    '\nThe number of down gene is ',nrow(DEG[DEG$change == 'DOWN',]))

MA <- ggplot(DEG_table, aes(x = log10(baseMean),y = log2FoldChange, color = change)) +
  geom_point(size = 1) +
  theme_set(theme_set(theme_bw(base_size = 12))) +
  theme_classic() +
  theme(legend.position="none") +
  geom_text_repel(data=DEG_table[which(DEG_table$change!="NOT"),], aes(x = log10(baseMean),y = log2FoldChange,label=SYMBOL),size=2) +
  geom_abline(intercept=c(-1,1),slope=0, colour="black",linetype = 'dashed') +
  xlab("log10 baseMean") + ylab("log2 FoldChange") +
  ggtitle(this_tile) + theme(plot.title = element_text(size=12,hjust=0.5)) +
  # scale_color_manual(values = c('darkgreen','grey','red'))
  scale_color_manual(values=c(NOT = "grey", UP = "red",DOWN="darkgreen")) 

MA
```

### 4.5 Volcano plot
```{r, fig.width=6, fig.height=6, fig.align = 'center'}
Volcano <- ggplot(DEG_table, aes(x = log2FoldChange,y = -log10(pvalue), color = change)) +
  geom_point(size = 1) +
  theme_set(theme_set(theme_bw(base_size = 12))) +
  theme_classic() +
  theme(legend.position="none") +
  geom_text_repel(data=DEG_table[which(DEG_table$change!="NOT"),], aes(x = log2FoldChange,y = -log10(pvalue),label=SYMBOL),size=2) +
  geom_vline(aes(xintercept=1),linetype="dashed") + 
  geom_vline(aes(xintercept=-1),linetype="dashed") + 
  # geom_abline(intercept=c(-1,1),slope=-1, colour="black") +
  xlab("log2 FoldChange") + ylab("-log10 p-value") +
  ggtitle(this_tile) + theme(plot.title = element_text(size=12,hjust=0.5)) +
  scale_color_manual(values=c(NOT = "grey", UP = "red",DOWN="darkgreen")) 

Volcano
```

### 4.6 Scatter plot
```{r, fig.width=6, fig.height=6, fig.align = 'center'}
Scatter <- ggplot(data=DEG_table, aes(x=log10(WT), y=log10(Mut),fill=change,color=change)) +
  geom_point(stat="identity",size=1) +
  geom_text_repel(data=DEG_table[which(DEG_table$change=='DOWN'),], aes(x=log10(WT), y=log10(Mut),label=SYMBOL),size=2,show.legend = FALSE) +
  geom_text_repel(data=DEG_table[which(DEG_table$change=='UP'),], aes(x=log10(WT), y=log10(Mut),label=SYMBOL),size=2,show.legend = FALSE) +
  ylab(paste0("log10 ",mut_name)) +
  xlab(paste0("log10 ",wt_name)) +
  theme_classic() +
  ggtitle(this_tile) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0)) +
  geom_abline(intercept = c(log10(2),log10(0.5)), slope = 1, linetype="dotted", colour="black") +
  geom_abline(intercept = 0, slope = 1 , colour="black") +
  scale_color_manual(values=c(NOT = "grey", UP = "red",DOWN="darkgreen")) 
Scatter
```

## 5 Different experssion gene Go Term
```{r}
up_gene <- filter(DEG_table,DEG_table$change == "UP")
up_gene <- na.omit(up_gene)
down_gene <- filter(DEG_table,DEG_table$change == "DOWN")
down_gene <- na.omit(down_gene)
```

### 5.1 UP gene Go Term --Molecular Function
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
up_MF <- enrichGO(gene = up_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "MF",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)

temp <- try(g <- dotplot(up_MF, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

### 5.2 UP gene Go Term --Cellular Component
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
up_CC <- enrichGO(gene = up_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "CC",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)

temp <- try(g <- dotplot(up_CC, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

### 5.3 UP gene Go Term --Biological Process
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
up_BP <- enrichGO(gene = up_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "BP",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)

temp <- try(g <- dotplot(up_BP, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

### 5.4 Down gene Go Term --Molecular Function
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
down_MF <- enrichGO(gene = down_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "MF",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)

temp <- try(g <- dotplot(down_MF, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

### 5.5 Down gene Go Term --Cellular Component
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
down_CC <- enrichGO(gene = down_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "CC",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)
temp <- try(g <- dotplot(down_CC, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

### 5.6 Down gene Go Term --Biological Process
```{r, fig.width=7, fig.height=6, fig.align = 'center'}
down_BP <- enrichGO(gene = down_gene$ENTREZID,
                 OrgDb = maize,
                 keyType = "ENTREZID",
                 ont = "BP",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = 0.1)

temp <- try(g <- dotplot(down_BP, showCategory = 6), silent = TRUE)
if('try-error' %in% class(temp)){g <- "No results found; Low degree of gene enrichment"}
temp <- try(print(g), silent = TRUE)
if('try-error' %in% class(temp)){print("No results found; Low degree of gene enrichment")}
```

**END**
